File: 005-一个复杂动画的实现/005-一个复杂动画的实现_OC/005-一个复杂动画的实现_OC/AnimationView/Layers/RectangleLayer.m
//
//  RectangleLayer.m
//  005-一个复杂动画的实现_OC
//
//  Created by 马遥 on 2018/1/23.
//  Copyright © 2018年 马遥. All rights reserved.
//

#import "RectangleLayer.h"

@interface RectangleLayer ()
@property (nonatomic, strong) UIBezierPath *rectangleFullPath;
@property (nonatomic, strong) UIColor *color;
@end

@implementation RectangleLayer

- (instancetype)init {
    self = [super init];
    if (self) {
        self.path = self.rectangleFullPath.CGPath;
        self.fillColor = [UIColor clearColor].CGColor;
        self.lineWidth = 5;
        _color = [UIColor colorWithRed:64/255.0 green:224/255.0 blue:176/255.0 alpha:1];
    }
    return self;
}

- (void)startStrokeWithColor:(UIColor *)color {
    if (color) {
        _color = color;
    }
    [self stroke];
}

- (void)stroke {
    [self addAnimation:[self strokeAnimation] forKey:nil];
}

#pragma mark - Animation

- (CABasicAnimation *)strokeAnimation {
    self.strokeColor = self.color.CGColor;
    CABasicAnimation *animation = [CABasicAnimation animationWithKeyPath:@"strokeEnd"];
    animation.fromValue = @0;
    animation.toValue = @1;
    animation.duration = 0.4;
    return animation;
}

#pragma mark - UIBezierPath

- (UIBezierPath *)rectangleFullPath {
    if (!_rectangleFullPath) {
        CGRect bounds = [UIScreen mainScreen].bounds;
        CGFloat width = (0.866 * (90/2.0+15) + 5/2.0) * 2;
        // 左下角
        CGPoint lowerLeftPoint = CGPointMake(bounds.size.width/2 - width/2, bounds.size.height/2 + 90/2.0/2.0 + 15/2 + 5/2.0);
        // 左上角
        CGPoint topLeftPoint = CGPointMake(lowerLeftPoint.x, lowerLeftPoint.y - width);
        // 右上角
        CGPoint topRightPoint = CGPointMake(lowerLeftPoint.x + width, topLeftPoint.y);
        // 右下角
        CGPoint lowerRightPoint = CGPointMake(topRightPoint.x, lowerLeftPoint.y);

        _rectangleFullPath = [UIBezierPath bezierPath];

        [_rectangleFullPath moveToPoint:lowerLeftPoint];
        [_rectangleFullPath addLineToPoint:topLeftPoint];
        [_rectangleFullPath addLineToPoint:topRightPoint];
        [_rectangleFullPath addLineToPoint:lowerRightPoint];
        [_rectangleFullPath closePath];
    }
    return _rectangleFullPath;
}

@end

File: 003-一个复杂动画的实现/004-一个复杂动画的实现_Swift/004-一个复杂动画的实现/Layers/RectangleLayer.swift
//
//  RectangleLayer.swift
//  004-一个复杂动画的实现
//
//  Created by 马遥 on 2018/1/24.
//  Copyright © 2018年 马遥. All rights reserved.
//

import UIKit

class RectangleLayer: CAShapeLayer {

    private lazy var rectangleFullPath: UIBezierPath = {
        let width: CGFloat = (0.866 * 60 + 2.5) * 2
        let lowerLeftPoint = CGPoint.init(x: Screen_Width / 2 - width / 2,
                                          y: Screen_Height / 2 + 45 / 2.0 + 10)
        let topLeftPoint = CGPoint.init(x: lowerLeftPoint.x,
                                        y: lowerLeftPoint.y - width)
        let topRightPoint = CGPoint.init(x: lowerLeftPoint.x + width,
                                         y: topLeftPoint.y)
        let lowerRightPoint = CGPoint.init(x: topRightPoint.x,
                                           y: lowerLeftPoint.y)
        
        let layer = UIBezierPath()
        layer.move(to: lowerLeftPoint)
        layer.addLine(to: topLeftPoint)
        layer.addLine(to: topRightPoint)
        layer.addLine(to: lowerRightPoint)
        layer.close()
        return layer
    }()
    
    override init() {
        super.init()
        fillColor = UIColor.clear.cgColor
        path = rectangleFullPath.cgPath
        lineWidth = 5
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    public func startStrokeWithColor(_ color: UIColor) {
        strokeColor = color.cgColor
        let animation = CABasicAnimation.init(keyPath: "strokeEnd")
        animation.fromValue = 0
        animation.toValue = 1
        animation.duration = 0.4
        self.add(animation, forKey: nil)
    }
    
}
